### ISSUE TITLE
Braintree dialogue disappear without callback when app minimised

### ISSUE THREAD

**Issue by HimanDhawan** [braintree/braintree_ios#893]
**Integration Details (please complete the following information):**
 - Braintree SDK Version: 4.38.0
 - Environment:Sandbox
 - iOS Version: Version 11+ 
 - Xcode Version: Version 14.0 
 - Device: iPhone 8
 - Integration type & version: 1.11.3
 - Development Processor Type: M1/ Apple Silicon

**Describe the bug**
In iOS, when we use Braintree BTPayPalDriver to create "requestBillingAgreement", dialogue appears. At this time if app is minimised, the dialogue disappears and none of "appContextWillSwitch" and "appContextDidReturn" is called.

**Code**
```
if let driver = self.payPalDriver {
                print("Setting up delegate")
                SDKStaticLoadingIndicatorWindow.showLoader()
                driver.appSwitchDelegate = self
                driver.viewControllerPresentingDelegate = self
                driver.requestBillingAgreement(req) { [weak self] (tokenizedPayPalAccount, error) in
                    
                }
            }
```
**To Reproduce**
1. Create BTPayPalDriver
2. Make driver.requestBillingAgreement request.
3. When dialogue appears, minimise the app.4. 
4. Again open the app and see dialogue is no longer present.5. 
5. No "appContextDidReturn" delegate is called.

**Expected behavior**
"appContextDidReturn" delegate method should be called.

**Screenshots**
![Simulator Screen Shot - iPhone 8 - 2022-11-15 at 14 00 27](https://user-images.githubusercontent.com/16226329/201868944-442a105c-86d3-41ea-b6fe-2189907cab73.png)

**Comment by jaxdesmarais** [comment]
Hello @HimanDhawan -

It is worth noting that version 4 of the SDK has been deprecated since February 2022 as noted in our [README](https://github.com/braintree/braintree_ios#versions) and [version 4.38.0](https://github.com/braintree/braintree_ios/blob/8964e496f38487439a70dc1d7fd0c1e9a91508f8/CHANGELOG.md#4380-2021-08-24) was released August 24th 2021. We highly recommend updating to version 5 of the SDK as the version you are using will be unsupported starting February 2023.

I've tested on a simulator running iOS 13 as well as iOS 12 but am unable to background the AlertController shown in the screenshot of the issue in either iOS version. I would recommend instead using `NotificationCenter` to hook into the `appMovedToBackground` selector. You can see an [example of adding an observer for this here](https://www.hackingwithswift.com/example-code/system/how-to-detect-when-your-app-moves-to-the-background). The SDK does not hook into a merchants App/SceneDelegate in any way so does not have a way of being notified in these cases and therefore our delegate cannot be called when the application is minimized.

**Comment by HimanDhawan** [comment]
@jaxdesmarais when you get the dialogue that I have shared above, lock the screen and comeback. It will not be there and we don't get any callback. Please check the video below.

https://user-images.githubusercontent.com/16226329/202646725-f05b0aaf-31ba-41a7-90ed-ab848bc05449.mov



Is this issue resolved in version 5 of Braintree?

**Comment by scannillo** [comment]
> In iOS, when we use Braintree BTPayPalDriver to create "requestBillingAgreement", dialogue appears. At this time if app is minimised, the dialogue disappears and none of "appContextWillSwitch" and "appContextDidReturn" is called.

To clarify, the Pay with PayPal flow does not ever leave your app. We present an [ASWebAuthenticationSession](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession?language=objc) which stays within the context of your application - so the app isn't moved to the background or inactive state at all. The app is always active. `appContextWillSwitch` and `appContextDidReturn` don't exist on iOS, but I am guessing you mean the [UIApplicationDelegate methods](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623032-application), such as `applicationDidBecomeActive(_:)`. It is intentional that these aren't called, as the PayPal flow is essentially just in a WebView presented within your app. 

To receive inspect success, error, or cancellation type results for the PayPal flow be sure to handle the `(result, error)` completion on `payPalDriver.tokenizePayPal()` [shown here in docs](https://developer.paypal.com/braintree/docs/guides/paypal/client-side#billing-address).

**Comment by scannillo** [comment]
> when you get the dialogue that I have shared above, lock the screen and comeback. It will not be there and we don't get any callback

Thank you for the additional detail, I am able to replicate this.

As for locking the phone while the alert is showing - the alert shown is an Apple system alert required for displaying `ASWebAuthenticationSession`s. 

[Here](https://github.com/braintree/braintree_ios/blob/b68b41623b5944f1320aa5aeabd825a1ff485a77/Sources/BraintreePayPal/BTPayPalDriver.m#L258-L291) is the part if our source code where we interface directly with `ASWebAuthenticationSession`. When locking and then unlocking the screen, the `ASWebAuthenticationSessionCompletionHandler` is never called by Apple. So there is no way for our SDK to surface any notification to you via our callback, because Apple simply doesn't handle this case.

In short - this looks like a limitation of Apple's `ASWebAuthenticationSession` and I'd encourage you to file an issue with them. Feel free to use this GitHub thread as a reference!

**Comment by scannillo** [comment]
> when you get the dialogue that I have shared above, lock the screen and comeback. It will not be there and we don't get any callback

Thank you for the additional detail, I am able to replicate this.

As for locking the phone while the alert is showing - the alert shown is an Apple system alert required for displaying `ASWebAuthenticationSession`s. 

[Here](https://github.com/braintree/braintree_ios/blob/b68b41623b5944f1320aa5aeabd825a1ff485a77/Sources/BraintreePayPal/BTPayPalDriver.m#L258-L291) is the part if our source code where we interface directly with `ASWebAuthenticationSession`. When locking and then unlocking the screen, the `ASWebAuthenticationSessionCompletionHandler` is never called by Apple. So there is no way for our SDK to surface any notification to you via our callback, because Apple simply doesn't handle this case.

In short - this looks like a limitation of Apple's `ASWebAuthenticationSession` and I'd encourage you to file an issue with them. Feel free to use this GitHub thread as a reference!

**Comment by scannillo** [comment]
Closing due to inactivity
